SQL Sync

NoahCarnahan opened the pull request:
Modify sync code to be backend agnostic. All tests in the `phone` app are passing.

TODO:
- ~~write sql functions instead of using django ORM~~
- ~~mark tests with `run_with_all_backends`~~
- ~~fix merge conflicts~~
- ~~remove notes / misc clean up~~
- ~~write tests for `CaseAccessors` API~~

can be reviewed commit-by-commit.
@snopoke @benrudolph @czue

NoahCarnahan committed:
Add get_case_ids_in_domain_by_owners() to CaseAccessors

NoahCarnahan committed:
Fix phone test app dependecies

NoahCarnahan committed:
Add get_open_case_ids() and get_closed_case_ids() to CaseAccessors

NoahCarnahan committed:
Use case_id property instead of _id

NoahCarnahan committed:
Use CaseAccessors for case lookup

NoahCarnahan committed:
Add get_extension_case_ids() to CaseAccessors

NoahCarnahan committed:
Add get_indexed_case_ids() to CaseAccessors

NoahCarnahan committed:
Add get_case_ids_modified_with_owner_since() to CaseAccessors

NoahCarnahan committed:
Add get_all_reverse_indices_info() to CaseAccessors

NoahCarnahan committed:
Clean up imports

NoahCarnahan committed:
Add get_last_modified_dates() to CaseAccessors

NoahCarnahan committed:
Add get_actions_for_form() to AbstractCommCareCase

NoahCarnahan committed:
Add modified_since_sync() and sync_log_id field to CaseTransaction model

NoahCarnahan committed:
fix assertion

NoahCarnahan committed:
Fix CommCareCaseSQL dynamic_case_properties()

NoahCarnahan committed:
Add run_with_all_backends decorators

dimagimon commented on corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py:
```diff
@@ -78,37 +71,37 @@ def get_payload(self):
                 if case.indices:
                     # and double check footprint for non-live cases
-                    extension_indices[case._id] = {index.identifier: index.referenced_id for index in case.indices
+                    extension_indices[case.case_id] = {index.identifier: index.referenced_id for index in case.indices
```
E501 line too long (118 > 115 characters)

dimagimon commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -328,6 +375,47 @@ def save_case(case):
+        raise NotImplementedError
+
+    @staticmethod
+    def get_extension_case_ids(domain, case_ids): # I need like this but without the relationship id check
```
E261 at least two spaces before inline comment

snopoke replied:
Have you run any of the tests with sql?

dimagimon commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -328,6 +375,47 @@ def save_case(case):
+    @staticmethod
+    def get_extension_case_ids(domain, case_ids): # I need like this but without the relationship id check
+        """
+        Given a base list of case ids, for those that are open, get all ids of all extension cases that reference them
```
E501 line too long (118 > 115 characters)

dimagimon commented on corehq/form_processor/interfaces/dbaccessors.py:
```diff
@@ -116,3 +149,29 @@ def get_case_xform_ids(self, case_ids):
+    def get_all_reverse_indices_info(self, case_ids):
+        return self.db_accessor.get_all_reverse_indices_info(self.domain, case_ids)
+
+CaseIndexInfo = namedtuple('CaseIndexInfo', ['case_id', 'identifier', 'referenced_id', 'referenced_type', 'relationship'])
```
E501 line too long (122 > 115 characters)

benrudolph commented:
nice! looking good :bullettrain_side:

snopoke commented on corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py:
```diff
@@ -166,13 +159,14 @@ def get_case_ids_for_owner(self, owner_id):
                 # for a clean owner's steady state sync, the base set is anything modified since last sync
-                return set(get_case_ids_modified_with_owner_since(
-                    self.restore_state.domain, owner_id, self.restore_state.last_sync_log.date
+                return set(CaseAccessors(domain).get_case_ids_modified_with_owner_since(
```
nit: could make CaseAccessors a variable to avoid creating it twice

NoahCarnahan replied:
Yeah, I'm thinking I'll store it as an instance variable of `CleanOwnerCaseSyncOperation`.

snopoke commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -221,6 +223,33 @@ def get_reverse_indices(case_id):
     @staticmethod
+    def get_all_reverse_indices_info(domain, case_ids):
+        query = CommCareCaseIndexSQL.objects.filter(
+            # domain=domain,  # TODO: Why isn't domain being saved on CommCareCaseIndexSQL objects?
```
no specific reason, you could join on case or else add it to CommCareCaseIndexSQL or trust that all case_ids are already in the correct domain.

snopoke commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -328,6 +375,47 @@ def save_case(case):
+            referenced_id__in=case_ids,
+            relationship_id=CommCareCaseIndexSQL.EXTENSION
+        )
+        # TODO: Filter by open cases?
```
Seems you need to join on the the case table anyway so you can add the domain check there

snopoke commented on corehq/form_processor/models.py:
```diff
@@ -455,7 +458,7 @@ def is_deleted(self):
 
     def dynamic_case_properties(self):
-        return self.case_json
+        return OrderedDict(sorted(self.case_json.iteritems()))
```
curious why this was necessary

NoahCarnahan replied:
`CommCareCase` [does it](https://github.com/dimagi/commcare-hq/blob/fd0a33742e65a1c19be383925a91d08521eb7e5e/corehq/ex-submodules/casexml/apps/case/models.py#L498), and one of the tests was failing without it because the dynamic properties in a restore payload were in a different order. A better solution might be to change the test to be order agnostic?

snopoke commented on corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py:
```diff
@@ -78,37 +71,37 @@ def get_payload(self):
-                [CommCareCase.wrap(doc) for doc in get_docs(CommCareCase.get_db(), ids)
-                 if not is_deleted(doc)]
+                [case for case in CaseAccessors(self.restore_state.domain).get_cases(ids)
+                 if not case.is_deleted]
```
Taking a closer look at this and the `case_needs_to_sync` function we may want to branch the logic here so that we can't take more advantage of the SQL db otherwise we're hitting the DB again for each case. Probably not in this PR but I think we could do the `modified_since_sync` check as part of the initial query.

snopoke replied:
Added [here](https://docs.google.com/document/d/1XX_WCEf8ozxt7cJoigu3R03W9vPqY45P7vFFY3stVO4/edit) so we don't forget about it.

NoahCarnahan replied:
thanks!

snopoke commented on corehq/form_processor/models.py:
```diff
@@ -524,6 +527,31 @@ def transactions(self):
+
+    def get_actions_for_form(self, xform):
+        from casexml.apps.case.xform import get_case_updates
+        # Will this import cause atrocious performance?
```
Just the import or the code below? I think it should be OK. If not we can probably memoize some of it on the form (since this gets called as part of form submission).

NoahCarnahan replied:
I was just thinking about the import statement (`from casexml.apps.case.xform import get_case_updates`). I wasn't sure how costly the import statement is, and if it would end up being a problem if this function is called many times.

snopoke replied:
This should be OK. It gets called once per case in the form (which is usually a small number). This is another thing that could be optimized though since its basically doing exactly the same work for each case (since the form will be the same). Fine for now though.

snopoke commented on corehq/form_processor/sql_functions/save_case_and_related_models.sql:
```diff
@@ -103,10 +103,10 @@ BEGIN
                 case_id, revoked, details
             ) VALUES (
-                case_transaction.form_id, case_transaction.server_date, case_transaction.type,
+                case_transaction.form_id, case_transaction.sync_log_id, case_transaction.server_date, case_transaction.type,
```
:+1:

snopoke commented:
not mergable but otherwise no blockers from me

czue commented:
looks good! seems like you have a lot of tests and proxy functions to write in your near future :)

NoahCarnahan committed:
Rename get_case_ids_in_domain_by_owners()

NoahCarnahan committed:
Temporarily disable SQL model db access restrictions

NoahCarnahan committed:
Move decorator

NoahCarnahan committed:
Merge remote-tracking branch 'origin/master' into sql-sync

Conflicts:
	corehq/form_processor/backends/sql/dbaccessors.py
	corehq/form_processor/models.py

snopoke commented:
not mergable

NoahCarnahan committed:
Fix botched merge

NoahCarnahan committed:
Use db functions instead of django ORM

NoahCarnahan committed:
revert removal of RestrictedManagers

dimagimon commented:
There are 5 errors:

* corehq/form_processor/backends/sql/dbaccessors.py, line 5 - F401 'Q' imported but unused
* corehq/form_processor/backends/sql/dbaccessors.py, line 261 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 335 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 433 - E261 at least two spaces before inline comment
* corehq/form_processor/backends/sql/dbaccessors.py, line 450 - E501 line too long (121 > 115 characters)

NoahCarnahan committed:
Make sql function PostgreSQL 9.1 compatible

dimagimon commented:
There are 5 errors:

* corehq/form_processor/backends/sql/dbaccessors.py, line 5 - F401 'Q' imported but unused
* corehq/form_processor/backends/sql/dbaccessors.py, line 261 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 335 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 433 - E261 at least two spaces before inline comment
* corehq/form_processor/backends/sql/dbaccessors.py, line 450 - E501 line too long (121 > 115 characters)

NoahCarnahan committed:
Add get_case_ids_modified_with_owner_since sql function

dimagimon commented:
There are 8 errors:

* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 88 - E501 line too long (118 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 5 - F401 'Q' imported but unused
* corehq/form_processor/backends/sql/dbaccessors.py, line 261 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 335 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 430 - E501 line too long (136 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 435 - E261 at least two spaces before inline comment
* corehq/form_processor/backends/sql/dbaccessors.py, line 452 - E501 line too long (121 > 115 characters)
* corehq/form_processor/interfaces/dbaccessors.py, line 177 - E501 line too long (122 > 115 characters)

NoahCarnahan committed:
Remove and fix comments

NoahCarnahan committed:
Fix incorrect doc string

NoahCarnahan commented:
@snopoke Now mergeable and updated to use pl/proxy functions.

NoahCarnahan commented:
@snopoke @czue @benrudolph (@gcapalbo might have some insight too)

Some of the db accessors require matching cases by `owner_id`, or `user_id` if the case has no `owner_id`. In the corresponding couch view functions, this is expressed with `var owner_id = doc.owner_id || doc.user_id;`. I've tried two different ways to express this in the postgres functions.

From `get_open_case_ids.sql`:
```SQL
      (
         -- owner_id matches
         case_table.owner_id = p_owner_id
         OR (
           -- owner_id is falsey and modified_by matches
           (case_table.owner_id is NULL OR case_table.owner_id = '') AND case_table.modified_by = p_owner_id
         )
      )
 ```
 
 and from `get_case_ids_modified_with_owner_since.sql`:
 ```SQL
 p_owner_id = COALESCE(NULLIF(case_table.owner_id, ''), NULLIF(case_table.modified_by, ''))
 ```
 
I think the second method is more readable. However, I think that postgres wouldn't be able to use `owner_id` or `modified_by` indexes when the columns are being used as arguments to functions like this. Can you confirm that postgres indeed would use indexes in the first method, but not in the second? I'm happy to spin up some dummy data and run an `EXPLAIN ANALYZE` query, but thought check to see if you know off the top of your head.

Another option would be to create an index on the expression itself...

dimagimon commented:
There are 7 errors:

* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 88 - E501 line too long (118 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 260 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 334 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 429 - E501 line too long (136 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 436 - E501 line too long (118 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 450 - E501 line too long (121 > 115 characters)
* corehq/form_processor/interfaces/dbaccessors.py, line 177 - E501 line too long (122 > 115 characters)

NoahCarnahan committed:
Avoid multiple creations of CaseAccessors object

dimagimon commented:
There are 8 errors:

* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 89 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 261 - F841 local variable 'domain' is assigned to but never used
* corehq/form_processor/backends/sql/dbaccessors.py, line 260 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 334 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 429 - E501 line too long (136 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 436 - E501 line too long (118 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 450 - E501 line too long (121 > 115 characters)
* corehq/form_processor/interfaces/dbaccessors.py, line 177 - E501 line too long (122 > 115 characters)

gcapalbo commented:
@NoahCarnahan Yeah if you reference the fields in a function you'll lose the advantage of the index, but you also won't gain much from the index in the first example either because that `OR` in there will still slow things down.

Is it true that every case should have an `owner_id` set and that the `owner_id || user_id` logic was there just as a lazy migration from when we didn't used to have case sharing? If so then I would actually vote that we make `owner_id` a `NOT NULL` column and set it to `owner_id || user_id` when we initially migrate the cases over, so that you can always join directly on `owner_id` (which would be much faster).

gcapalbo commented:
@NoahCarnahan Also is `modified_by` a typo there - should that be `user_id`?

NoahCarnahan commented:
Thanks @gcapalbo. I'm not sure if the `owner_id || user_id` logic is actually just an artifact of a lazy migration or not, but it would be nice if we could handle it in the migration. @czue do you know?

Thanks for pointing out the `modified_by` discrepancy. I forgot to mention in the original comment, but `modified_by` is correct. The `user_id` field has been renamed to `modified_by` in the sql model.

NoahCarnahan committed:
Add basic tests for new CaseAccessorSQL methods

dimagimon commented:
There are 13 errors:

* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 89 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 261 - F841 local variable 'domain' is assigned to but never used
* corehq/form_processor/backends/sql/dbaccessors.py, line 260 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 334 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 429 - E501 line too long (136 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 450 - E501 line too long (121 > 115 characters)
* corehq/form_processor/interfaces/dbaccessors.py, line 177 - E501 line too long (122 > 115 characters)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 369 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 378 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 385 - F841 local variable 'case1' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 387 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 429 - E303 too many blank lines (2)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 495 - F841 local variable 'case3' is assigned to but never used

NoahCarnahan committed:
Add one more test for CaseAccessorSQL

dimagimon commented:
There are 13 errors:

* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 89 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 261 - F841 local variable 'domain' is assigned to but never used
* corehq/form_processor/backends/sql/dbaccessors.py, line 260 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 334 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 429 - E501 line too long (136 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 450 - E501 line too long (121 > 115 characters)
* corehq/form_processor/interfaces/dbaccessors.py, line 177 - E501 line too long (122 > 115 characters)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 369 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 378 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 385 - F841 local variable 'case1' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 387 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 429 - E303 too many blank lines (2)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 495 - F841 local variable 'case3' is assigned to but never used

NoahCarnahan committed:
Improve get_case_ids_modified_with_owner_since sql function

Replace COALESCE and NULLIF functions with AND and OR statements
so that indexes can be used.

dimagimon commented:
There are 13 errors:

* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 89 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 261 - F841 local variable 'domain' is assigned to but never used
* corehq/form_processor/backends/sql/dbaccessors.py, line 260 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 334 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 429 - E501 line too long (136 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 450 - E501 line too long (121 > 115 characters)
* corehq/form_processor/interfaces/dbaccessors.py, line 177 - E501 line too long (122 > 115 characters)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 369 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 378 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 385 - F841 local variable 'case1' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 387 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 429 - E303 too many blank lines (2)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 495 - F841 local variable 'case3' is assigned to but never used

NoahCarnahan commented:
Updated owner filter to use `AND` and `OR` instead of `COALESCE` and `NULLIF`. Also added tests. I think this is good to merge.

gcapalbo commented on corehq/sql_accessors/sql_templates/case_has_transactions_since_sync.sql:
```diff
@@ -0,0 +1,11 @@
+      SELECT TRUE FROM form_processor_casetransaction AS transaction_table
+        WHERE transaction_table.case_id = $1
+          AND transaction_table.server_date > $3
+          AND transaction_table.sync_log_id IS DISTINCT FROM $2
```
(not a blocker) I think this would be slightly faster if you added `LIMIT 1`

gcapalbo commented on corehq/sql_accessors/sql_templates/get_all_reverse_indices.sql:
```diff
@@ -0,0 +1,12 @@
+    SELECT * FROM form_processor_commcarecaseindexsql AS case_index_table
+    WHERE
+      -- case_index_table.domain = domain_name AND -- TODO: Uncomment if CaseIndex ever uses its domain column
+      case_index_table.referenced_id = ANY(case_ids)
```
Do we have a sense for the max number of case_ids that will be passed into this function? `ANY` is essentially a long `OR` statement, and `OR`s tend to be slow.

I don't think we should do this now (since it also increases complexity), but just to keep in mind - a faster way to do this kind of thing is to setup a temporary table with an indexed column for case_id, insert the case_ids into it, and then use a join to the temporary table to filter.  But that's overkill if the number of case_ids will always be small (i.e., < 10)

snopoke replied:
Not saying that method wouldn't be better for large lists but PG > 9.4 has made some good optimisations to IN / ANY queries as well so probably not worth doing anything unless we see that this performs poorly.

gcapalbo replied:
yup, agreed, was just mentioning as something to keep in mind in case we find ourselves checking big lists at some point and it performs poorly

gcapalbo commented on corehq/sql_accessors/sql_templates/get_case_ids_modified_with_owner_since.sql:
```diff
@@ -0,0 +1,19 @@
+      -- AND owner_id matches OR (owner_id is falsey AND modified_by matches)
+      AND (
+         case_table.owner_id = p_owner_id
+         OR (
```
This is probably ok for now, but I think we definitely want to address updating this in another PR.  I think this `OR` here with reference to `modified_by` will basically make us lose the advantage of the index on `owner_id`, so if we just set `owner_id` appropriately on the initial case migration we can join exclusively on `owner_id` and that would be a lot faster.

gcapalbo commented on corehq/sql_accessors/sql_templates/get_closed_case_ids.sql:
```diff
@@ -0,0 +1,13 @@
+    WHERE
+      case_table.closed = TRUE
+      AND case_table.domain = domain_name
+      AND case_table.owner_id = p_owner_id
```
Did this one need to check `modified_by` in case `owner_id` is blank? (Though no need to update if we can confirm that we'll set `owner_id` appropriately on initial migration)

gcapalbo commented on corehq/sql_accessors/sql_templates/get_open_case_ids.sql:
```diff
@@ -0,0 +1,19 @@
+      AND (
+        -- owner_id matches
+        case_table.owner_id = p_owner_id
+        OR (
```
(same comment as above regarding `owner_id`)

gcapalbo commented:
@NoahCarnahan Made a few comments but looks good to me.  I think I'd feel more comfortable giving @snopoke or @czue a chance to sign off on this one when they get back though since they have the most context.  Are there any dependencies on this getting merged soon or can it wait until then?

czue commented:
@NoahCarnahan I haven't had a chance to review this yet (and not sure when I will) but I think it would be totally fine to ditch the `user_id` fallback and always assume/ensure `owner_id` is set properly. That's been a confusing wart in the code for ages. So I'd second @gcapalbo's suggestion of making that not null and not worrying about `user_id` at all.

proteusvacuum commented:
This probably isn't important, but thought I'd mention anyway: Extension Cases can have an owner_id set to '-' which indicates that they are unowned extensions. Not sure if this will become a performance issue anywhere (haven't looked closely at the SQL functions that pull the cases). As long as this case is handled, I don't think there should be an issue otherwise (and looks like all tests are passing, so looks good anyway).

snopoke commented:
+100 on making owner_id not null! Looking good @NoahCarnahan pending comments. Are you running the tests locally in both sharded and non-sharded setup? Travis only runs non-sharded so doesn't test the pl-proxy functions. Example settings in `localsettings.example.py`. To run sharded set `USE_PARTITIONED_DATABASE` to `True`: https://github.com/dimagi/commcare-hq/blob/master/localsettings.example.py#L18

NoahCarnahan commented:
@gcapalbo yeah it's fine to wait on snopoke or czue to merge, this isn't blocking anything. Thanks for the feedback.

NoahCarnahan commented:
Thanks for the warning @proteusvacuum, I'll keep that in mind in case any issues arise.

NoahCarnahan committed:
Improve readability of case_has_transactions_since_sync

NoahCarnahan committed:
Add not null constraint to case.owner_id

NoahCarnahan committed:
Fix typos in plproxy functions

NoahCarnahan committed:
Rename sql function

dimagimon commented:
There are 13 errors:

* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 89 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 261 - F841 local variable 'domain' is assigned to but never used
* corehq/form_processor/backends/sql/dbaccessors.py, line 260 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 334 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 429 - E501 line too long (136 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 450 - E501 line too long (126 > 115 characters)
* corehq/form_processor/interfaces/dbaccessors.py, line 177 - E501 line too long (122 > 115 characters)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 369 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 378 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 385 - F841 local variable 'case1' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 387 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 429 - E303 too many blank lines (2)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 495 - F841 local variable 'case3' is assigned to but never used

NoahCarnahan commented:
Added a `user_id` is not null constraint, and updated the sql functions to only query on `owner_id`.

@snopoke thanks for pointing out the `USE_PARTITIONED_DATABASE` setting, using that locally now and all tests in the `phone` app passing.

gcapalbo commented:
:+1: from my side, will give @snopoke or @czue a chance to sign off when they're back

czue commented:
I'll defer to @snopoke on this but no blockers from me

snopoke commented:
Just the one thing outstanding re setting owner_id to user_id if it's not explicitly set.

NoahCarnahan committed:
Use user_id as owner_id if none given in case create action

dimagimon commented:
There are 13 errors:

* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 89 - E501 line too long (118 > 115 characters)
* corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py, line 261 - F841 local variable 'domain' is assigned to but never used
* corehq/form_processor/backends/sql/dbaccessors.py, line 260 - E501 line too long (117 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 334 - E501 line too long (120 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 429 - E501 line too long (136 > 115 characters)
* corehq/form_processor/backends/sql/dbaccessors.py, line 450 - E501 line too long (126 > 115 characters)
* corehq/form_processor/interfaces/dbaccessors.py, line 177 - E501 line too long (122 > 115 characters)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 369 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 378 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 385 - F841 local variable 'case1' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 387 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 429 - E303 too many blank lines (2)
* corehq/form_processor/tests/test_case_dbaccessor.py, line 495 - F841 local variable 'case3' is assigned to but never used

snopoke commented:
@NoahCarnahan GTG from my side. Let me know if you want to get rid of any of the lint errors before merging.

NoahCarnahan committed:
lint

dimagimon commented:
There are 5 errors:

* corehq/form_processor/tests/test_case_dbaccessor.py, line 369 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 378 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 385 - F841 local variable 'case1' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 387 - F841 local variable 'case3' is assigned to but never used
* corehq/form_processor/tests/test_case_dbaccessor.py, line 494 - F841 local variable 'case3' is assigned to but never used

NoahCarnahan commented:
@snopoke removed a few of the lint errors. I'm going to keep around the unused local variables in the tests because they are potentially useful if we want to make the tests more comprehensive. Happy to remove them though if people think it makes things messy.
GTG from my perspective.

NoahCarnahan commented:
self merging because I believe simon is offline for the week

czue commented:
:boom:

gcapalbo commented:
does any special migration need to happen to set a sql case's owner_id to be `owner_id || user_id` for old cases?

czue commented:
@gcapalbo yes, but we aren't worrying about migrating existing data for a long time. it does seem worth noting that requirement somewhere though. @snopoke any thoughts where it belongs?

snopoke commented:
I've added a 'migration' section to the doc we've been using: https://docs.google.com/document/d/1XX_WCEf8ozxt7cJoigu3R03W9vPqY45P7vFFY3stVO4/edit
